localstack/config.py

```diff
@@ -32,6 +32,115 @@
 load_start_time = time.time()
 
 
+class Directories:
+    """
+    Holds the different directories available to localstack. Some directories are shared between the host and the
+    localstack container, some live only on the host and some only in the container.
+
+    Attributes:
+        infra:      container only; static infrastructure: binaries and libraries packaged with the image
+        var:        shared; variable data computed at runtime: lazy-loaded binaries, ssl cert, ...
+        cache:      shared; ephemeral data that has to persist across localstack runs and reboots
+        tmp:        individual; ephemeral data that does not have to persist across localstack runs but not reboots
+        shared_tmp: shared; ephemeral data used to communicate between container<->lambdas
+        data:       shared; holds localstack state, pods, ...
+        config:     host only; pre-defined configuration values, cached credentials, machine id, ...
+        init:       shared; user-defined provisioning scripts executed when in the container when it starts
+        logs:       shared; log files produced by localstack
+    """
+
+    infra: str
+    var: str
+    cache: str
+    tmp: str
+    shared_tmp: str
+    data: str
+    config: str
+    init: str
+    logs: str
+
+    def __init__(
+        self,
+        infra: str = None,
+        var: str = None,
+        cache: str = None,
+        tmp: str = None,
+        shared_tmp: str = None,
+        data: str = None,
+        config: str = None,
+        init: str = None,
+        logs: str = None,
+    ) -> None:
+        super().__init__()
+        self.infra = infra
+        self.var = var
+        self.cache = cache
+        self.tmp = tmp
+        self.shared_tmp = shared_tmp
+        self.data = data
+        self.config = config
+        self.init = init
+        self.logs = logs
+
+    @staticmethod
+    def from_config():
+        """Returns Localstack directory paths from the config/environment variables defined by the config."""
+        return Directories(
+            infra=INSTALL_DIR_INFRA,
+            var=TMP_FOLDER,  # TODO: add variable
+            cache=TMP_FOLDER,  # TODO: add variable
+            tmp=TMP_FOLDER,  # TODO: move default value to /tmp/localstack/host
+            shared_tmp=HOST_TMP_FOLDER,  # TODO: move default value to /tmp/localstack/shared
+            data=DATA_DIR,
+            config=None,  # TODO: will be introduced once .localstack config file has been refactored
+            init=None,  # TODO: introduce environment variable
+            logs=TMP_FOLDER,  # TODO: move to /tmp/localstack/logs
+        )
+
+    @staticmethod
+    def for_container() -> "Directories":
+        """
+        /var/lib/localstack. Returns Localstack directory paths as they are defined within the container. Everything
+        shared and writable lives in /var/lib/localstack or /tmp/localstack.
```

Comment:
The directory structure with `/var` and `/etc` would be problematic when we consider moving to a more permissive user configuration (i.e. using a specific lower-priviledged / non-root user), since localstack itself would not be allowed to write there anymore.  Should we move this to the home directory (of the user running localstack in host-mode, or of the `localstack` user in the container)? On the other hand, considering the nature of LocalStack with all its low-level services (DNS server,...) it might be really challenging to restrict its permissions while supporting the full feature-set.

Reply:
i may have done a poor job communicating there: /var/ and /etc will only be used inside the container. the local mount would be, e.g. ~/.cache/localstack -> /var/lib/localstack  ~/.localstack -> /etc/localstack (ro)

Reply:
Yeah, I know, I was thinking about restricting the permissions of localstack in the container itself. For example, OpenShift clusters do not allow any root uid usage (without an explicit security context constraints configured).